Add symmetric int16 support to tflu softmax reference kernel

sicong-li-arm opened the pull request:
Signed-off-by: SiCongLi <sicong.li@arm.com>

gbaned commented:
@sicong-li-arm Can you please resolve conflicts? Thanks!

sicong-li-arm commented:
> @sicong-li-arm Can you please resolve conflicts? Thanks!

Done. Thanks for the reminder!

sicong-li-arm committed:
Add symmetric int16 support to tflu softmax reference kernel

sicong-li-arm commented:
Used placement new instead of cast to properly construct SoftmaxOpData. Removed the refactoring code to simplify the commit.

sicong-li-arm commented:
@hajuho @advaitjain Could you please have a look into this patch? Thanks!

hajuho commented:
@advaitjain I'm not used with tfl micro code. Could you review this change?

sicong-li-arm commented:
Hi @advaitjain could you please have a look into this PR? Much appreciated!

petewarden commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -29,22 +29,39 @@ namespace micro {
+// Softmax parameter data that persists in user_data
+struct SoftmaxOpData {
+  struct SoftmaxParams params = {};
+  float table[256]{};
```
Can you show me where this is used? I did a quick code search, and I don't see it being accessed.

sicong-li-arm replied:
Done removed.
It was originally intended for LUT for int8 but was not used.

petewarden commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -29,22 +29,39 @@ namespace micro {
+struct SoftmaxOpData {
+  struct SoftmaxParams params = {};
+  float table[256]{};
+  const int size_of_lut = 513;
```
Can you make this constexpr instead of const?

sicong-li-arm replied:
Done

petewarden commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -72,12 +99,46 @@ TfLiteStatus CalculateSoftmaxParams(TfLiteContext* context,
 
+void* SoftmaxInit(TfLiteContext* context, const char* buffer, size_t length) {
+  void* raw = nullptr;
+  context->AllocatePersistentBuffer(context, sizeof(SoftmaxOpData), &raw);
```
Could you break this allocation up so that exp_lut and one_over_one_plus_x_lut are allocated separately, and only if the type is int16? Otherwise the tables will take up 2KB of RAM even for types that aren't using them. Something like

```
struct SoftmaxOpData {
...
int16_t exp_lut*;
int16_t one_over_one_plus_x_lut*;
};
...
  void* raw_op_data = nullptr;
  context->AllocatePersistentBuffer(context, sizeof(SoftmaxOpData), &raw_op_data);
  SoftmaxOpData* op_data = reinterpret_cast<SoftmaxOpData*>(raw_op_data);
  if (input->type == kTfLiteInt16) {
    void* raw_exp_lut;
    context->AllocatePersistentBuffer(context, sizeof(int16_t) * size_of_lut, &raw_exp_lut);
    op_data->exp_lut = reinterpret_cast<int16_t*>(raw_exp_lut);
   ...
```

sicong-li-arm replied:
Done. That's a good point!

petewarden commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -72,12 +99,46 @@ TfLiteStatus CalculateSoftmaxParams(TfLiteContext* context,
+    data->params.exp_lut = data->exp_lut;
+    // exp LUT only used on negative values
+    // we consider exp(-10.0) is insignificant to accumulation
+    gen_lut([](double value) { return std::exp(value); }, -10.0, 0.0,
```
Do these have to be doubles, or would floats work? Most of our platforms will fall back to software emulation for doubles, which can be slow, even just for initializations.

sicong-li-arm replied:
Done changed to float.

petewarden commented on tensorflow/lite/micro/testing/test_utils.h:
```diff
@@ -29,6 +29,9 @@ limitations under the License.
 namespace testing {
 
+// Float tolerance for quantized int16 tests
+constexpr float kQuantizedToleranceInt16 = 2 * (1. / 4096);
```
Can you explain the reasoning behind this value in the comment?

sicong-li-arm replied:
It was copied from other int16 tests but you are right to point out that it's arbitrary and I noticed this tolerance was a bit too relaxed on some output values. I changed it to relative tolerance instead as I think it's more robust and meaningful.

petewarden reviewed (dismissed):
Thanks for this! Could you also link to the design doc or any RFC we have for the int16 integration, since before now it hasn't been used on the Micro side and I'm not as familiar with how we're approaching it.

sicong-li-arm commented:
> Thanks for this! Could you also link to the design doc or any RFC we have for the int16 integration, since before now it hasn't been used on the Micro side and I'm not as familiar with how we're approaching it.

@petewarden  Thanks so much for your review! I'm don't have any generic design doc for int16 integration but I can ask around. As for this particular integration, I was simply integrating the new LUT-based tflite int16 softmax routine, and reusing its logic for generating the LUTs. The original tflite patch is [here](https://github.com/tensorflow/tensorflow/pull/35996)

sicong-li-arm commented:
Hi @petewarden! I've addressed your comments. Could you please have a look? Thanks very much! :)

rthadur commented:
@sicong-li-arm please check below errors 

`tensorflow/lite/micro/testing/test_utils.cc:16:
.//tensorflow/lite/micro/testing/test_utils.h:97:19: error: unused variable 'input_dims' [-Werror,-Wunused-variable]
 TfLiteIntArray* input_dims = IntArrayFromInitializer(input_dims_data);
                 ^
1 error generated.
Broken by missing target ///tensorflow/lite/micro/kernels:micro_ops
content_copy
/tensorflow/lite/micro/kernels/softmax.cc:55:7: error: implicit conversion increases floating-point precision: 'float' to 'double' 
     TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.//tensorflow/lite/c/common.h:231:9: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   if (delta > epsilon) {                                             \
       ^~~~~ ~
/tensorflow/lite/micro/kernels/softmax.cc:56:36: error: implicit conversion increases floating-point precision: 'float' to 'double' 
                         (0.001 * 1.f / 32768));
                                ~ ^~~
.//tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   if (delta > epsilon) {                                             \
               ^~~~~~~
/tensorflow/lite/micro/kernels/softmax.cc:55:62: error: implicit conversion increases floating-point precision: 'float' to 'double' 
     TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
.//tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
                        __FILE__, __LINE__, #a, #b, (a), (b));        \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
.//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
   (context)->ReportError((context), __VA_ARGS__); \
   ~                                 ^~~~~~~~~~~
/tensorflow/lite/micro/kernels/softmax.cc:55:51: error: implicit conversion increases floating-point precision: 'float' to 'double' 
     TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
.//tensorflow/lite/c/common.h:233:55: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
                        __FILE__, __LINE__, #a, #b, (a), (b));        \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
.//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
   (context)->ReportError((context), __VA_ARGS__); \
   ~                                 ^~~~~~~~~~~
/tensorflow/lite/micro/kernels/softmax.cc:61:9: error: implicit conversion increases floating-point precision: 'float' to 'double' 
       TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.//tensorflow/lite/c/common.h:231:9: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   if (delta > epsilon) {                                             \
       ^~~~~ ~
/tensorflow/lite/micro/kernels/softmax.cc:62:38: error: implicit conversion increases floating-point precision: 'float' to 'double' 
                           (0.001 * 1.f / 65536));
                                  ~ ^~~
.//tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   if (delta > epsilon) {                                             \
               ^~~~~~~
/tensorflow/lite/micro/kernels/softmax.cc:61:64: error: implicit conversion increases floating-point precision: 'float' to 'double' 
       TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
.//tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
                        __FILE__, __LINE__, #a, #b, (a), (b));        \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
.//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
   (context)->ReportError((context), __VA_ARGS__); \
   ~                                 ^~~~~~~~~~~
/tensorflow/lite/micro/kernels/softmax.cc:61:53: error: implicit conversion increases floating-point precision: 'float' to 'double' 
       TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
.//tensorflow/lite/c/common.h:233:55: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
                        __FILE__, __LINE__, #a, #b, (a), (b));        \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
.//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
   (context)->ReportError((context), __VA_ARGS__); \
   ~                                 ^~~~~~~~~~~
/tensorflow/lite/micro/kernels/softmax.cc:76:31: error: implicit conversion increases floating-point precision: 'float' to 'double' 
         input->params.scale * params->beta /
         ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~ ~
/tensorflow/lite/micro/kernels/softmax.cc:152:51: error: implicit conversion increases floating-point precision: 'float' to 'double' 
   gen_lut([](float value) { return 1.0 / (1.0 + value); }, 0.0, 1.0,
                                               ~ ^~~~~
/tensorflow/lite/micro/kernels/softmax.cc:130:9: error: unused variable 'params' [-Werror,-Wunused-variable]
 auto* params = reinterpret_cast<TfLiteSoftmaxParams*>(node->builtin_data);

`

sicong-li-arm commented:
> @sicong-li-arm please check below errors
> 
> `tensorflow/lite/micro/testing/test_utils.cc:16:
> .//tensorflow/lite/micro/testing/test_utils.h:97:19: error: unused variable 'input_dims' [-Werror,-Wunused-variable]
> TfLiteIntArray* input_dims = IntArrayFromInitializer(input_dims_data);
> ^
> 1 error generated.
> Broken by missing target ///tensorflow/lite/micro/kernels:micro_ops
> content_copy
> /tensorflow/lite/micro/kernels/softmax.cc:55:7: error: implicit conversion increases floating-point precision: 'float' to 'double'
> TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
> ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> .//tensorflow/lite/c/common.h:231:9: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> if (delta > epsilon) {
> ^~~~~ ~
> /tensorflow/lite/micro/kernels/softmax.cc:56:36: error: implicit conversion increases floating-point precision: 'float' to 'double'
> (0.001 * 1.f / 32768));
> ~ ^~~
> .//tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> if (delta > epsilon) {
> ^~~~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:55:62: error: implicit conversion increases floating-point precision: 'float' to 'double'
> TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
> .//tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> **FILE**, **LINE**, #a, #b, (a), (b));
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
> .//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
> (context)->ReportError((context), **VA_ARGS**);
> ~ ^~~~~~~~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:55:51: error: implicit conversion increases floating-point precision: 'float' to 'double'
> TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 32768,
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
> .//tensorflow/lite/c/common.h:233:55: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> **FILE**, **LINE**, #a, #b, (a), (b));
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
> .//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
> (context)->ReportError((context), **VA_ARGS**);
> ~ ^~~~~~~~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:61:9: error: implicit conversion increases floating-point precision: 'float' to 'double'
> TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
> ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> .//tensorflow/lite/c/common.h:231:9: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> if (delta > epsilon) {
> ^~~~~ ~
> /tensorflow/lite/micro/kernels/softmax.cc:62:38: error: implicit conversion increases floating-point precision: 'float' to 'double'
> (0.001 * 1.f / 65536));
> ~ ^~~
> .//tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> if (delta > epsilon) {
> ^~~~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:61:64: error: implicit conversion increases floating-point precision: 'float' to 'double'
> TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
> .//tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> **FILE**, **LINE**, #a, #b, (a), (b));
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
> .//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
> (context)->ReportError((context), **VA_ARGS**);
> ~ ^~~~~~~~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:61:53: error: implicit conversion increases floating-point precision: 'float' to 'double'
> TF_LITE_ENSURE_NEAR(context, output->params.scale, 1.f / 65536,
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~
> .//tensorflow/lite/c/common.h:233:55: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> **FILE**, **LINE**, #a, #b, (a), (b));
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~
> .//tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
> (context)->ReportError((context), **VA_ARGS**);
> ~ ^~~~~~~~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:76:31: error: implicit conversion increases floating-point precision: 'float' to 'double'
> input->params.scale * params->beta /
> ~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~ ~
> /tensorflow/lite/micro/kernels/softmax.cc:152:51: error: implicit conversion increases floating-point precision: 'float' to 'double'
> gen_lut([](float value) { return 1.0 / (1.0 + value); }, 0.0, 1.0,
> ~ ^~~~~
> /tensorflow/lite/micro/kernels/softmax.cc:130:9: error: unused variable 'params' [-Werror,-Wunused-variable]
> auto* params = reinterpret_cast<TfLiteSoftmaxParams*>(node->builtin_data);
> 
> `

Hi @rthadur. Thanks for posting this. I forgot to turn on -Wall when building. I'll fix these soon.

rthadur reviewed (dismissed):
Thank you

rthadur commented:
@sicong-li-arm can you please resolve conflicts.

sicong-li-arm commented:
> @sicong-li-arm can you please resolve conflicts.

@rthadur Done. Thanks for the reminder!

sicong-li-arm commented:
Hi @rthadur. Could you please have a look once you're available? Thanks! :)

rthadur commented:
@sicong-li-arm can you please check below errors ?

`tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                         1.f / 32768, (0.001 * 1.f / 32768));
                         ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:230:26: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
                     ~  ^
tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                         1.f / 32768, (0.001 * 1.f / 32768));
                         ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:230:40: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
                                   ~  ^
tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                         1.f / 32768, (0.001 * 1.f / 32768));
                         ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:230:48: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
                                              ^  ~
tensorflow/lite/micro/kernels/softmax.cc:56:49: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                         1.f / 32768, (0.001 * 1.f / 32768));
                                             ~ ^~~
./tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   if (delta > epsilon) {                                             \
               ^~~~~~~
tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                         1.f / 32768, (0.001 * 1.f / 32768));
                         ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
                        __FILE__, __LINE__, #a, #b, (a), (b));        \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
./tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
   (context)->ReportError((context), __VA_ARGS__); \
   ~                                 ^~~~~~~~~~~
tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                           1.f / 65536, (0.001 * 1.f / 65536));
                           ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:230:26: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
                     ~  ^
tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                           1.f / 65536, (0.001 * 1.f / 65536));
                           ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:230:40: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
                                   ~  ^
tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                           1.f / 65536, (0.001 * 1.f / 65536));
                           ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:230:48: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
                                              ^  ~
tensorflow/lite/micro/kernels/softmax.cc:62:51: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                           1.f / 65536, (0.001 * 1.f / 65536));
                                               ~ ^~~
./tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
   if (delta > epsilon) {                                             \
               ^~~~~~~
tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
                           1.f / 65536, (0.001 * 1.f / 65536));
                           ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
                        __FILE__, __LINE__, #a, #b, (a), (b));        \
                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
./tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
   (context)->ReportError((context), __VA_ARGS__); \
   ~                                 ^~~~~~~~~~~
tensorflow/lite/micro/kernels/softmax.cc:152:51: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
   gen_lut([](float value) { return 1.0 / (1.0 + value); }, 0.0, 1.0,

`

sicong-li-arm commented:
> @sicong-li-arm can you please check below errors ?
> 
> `tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 32768, (0.001 * 1.f / 32768));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:230:26: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));
> ~ ^
> tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 32768, (0.001 * 1.f / 32768));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:230:40: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));
> ~ ^
> tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 32768, (0.001 * 1.f / 32768));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:230:48: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));
> ^ ~
> tensorflow/lite/micro/kernels/softmax.cc:56:49: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 32768, (0.001 * 1.f / 32768));
> ~ ^~~
> ./tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> if (delta > epsilon) {
> ^~~~~~~
> tensorflow/lite/micro/kernels/softmax.cc:56:31: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 32768, (0.001 * 1.f / 32768));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> **FILE**, **LINE**, #a, #b, (a), (b));
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
> ./tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
> (context)->ReportError((context), **VA_ARGS**);
> ~ ^~~~~~~~~~~
> tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 65536, (0.001 * 1.f / 65536));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:230:26: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));
> ~ ^
> tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 65536, (0.001 * 1.f / 65536));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:230:40: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));
> ~ ^
> tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 65536, (0.001 * 1.f / 65536));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:230:48: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));
> ^ ~
> tensorflow/lite/micro/kernels/softmax.cc:62:51: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 65536, (0.001 * 1.f / 65536));
> ~ ^~~
> ./tensorflow/lite/c/common.h:231:17: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> if (delta > epsilon) {
> ^~~~~~~
> tensorflow/lite/micro/kernels/softmax.cc:62:33: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> 1.f / 65536, (0.001 * 1.f / 65536));
> ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./tensorflow/lite/c/common.h:233:60: note: expanded from macro 'TF_LITE_ENSURE_NEAR'
> **FILE**, **LINE**, #a, #b, (a), (b));
> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~
> ./tensorflow/lite/c/common.h:161:39: note: expanded from macro 'TF_LITE_KERNEL_LOG'
> (context)->ReportError((context), **VA_ARGS**);
> ~ ^~~~~~~~~~~
> tensorflow/lite/micro/kernels/softmax.cc:152:51: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> gen_lut([](float value) { return 1.0 / (1.0 + value); }, 0.0, 1.0,
> 
> `

Hi @rthadur thantks for posting this. I wonder if the -Wdouble-promotion is strictly required? The implicit promotion to double seems to be required by the printf style format from the error reporter. I could explicitly cast them to double of course, but the implicit promotion seems to be everywhere within the codebase?

advaitjain commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -72,6 +101,68 @@ TfLiteStatus CalculateSoftmaxParams(TfLiteContext* context,
+  TfLiteTensor* output = GetOutput(context, node, 0);
+
+  TFLITE_DCHECK(node->user_data != nullptr);
+  SoftmaxOpData* data = new (node->user_data) SoftmaxOpData;
```
this should be a static_cast. If you rebase then an internal change has already changed this but you may have to resolve the conflicts.

sicong-li-arm replied:
Done.
Changed to static_cast.

advaitjain reviewed (changes requested):
This change is failing internal tests due to additional cflags that are applied internally but not on the OSS builds:
-Wdouble-promotion -Werror

WIth these tow flags added to CXXFLAGS in micro/tools/make/Makefile, I can reproduce the errors that we are seeing internally.

That the flags are different for the internal and external builds is a bug and we have a change that is being reviewed that should fix that issue. In the meantime, if you could rebase, resolve conflicts and then locally apply the additional cflags, you can likely get this change to a state that it passes internal and external presubmit checks.

sicong-li-arm commented:
> This change is failing internal tests due to additional cflags that are applied internally but not on the OSS builds:
> -Wdouble-promotion -Werror
> 
> WIth these tow flags added to CXXFLAGS in micro/tools/make/Makefile, I can reproduce the errors that we are seeing internally.
> 
> That the flags are different for the internal and external builds is a bug and we have a change that is being reviewed that should fix that issue. In the meantime, if you could rebase, resolve conflicts and then locally apply the additional cflags, you can likely get this change to a state that it passes internal and external presubmit checks.

Hi @advaitjain thanks for the review! I have pushed fixes for the double promotion.

advaitjain commented on tensorflow/lite/c/common.h:
```diff
@@ -225,6 +225,16 @@ void TfLiteFloatArrayFree(TfLiteFloatArray* a);
+  do {                                                                 \
+    auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
+    if (delta > epsilon) {                                             \
+      TF_LITE_KERNEL_LOG((context), "%s:%d %s not near %s (%d != %d)", \
```
change the format specifiers to be %f? its a limitation of macros, but we only really will use _NEAR with floats and doubles, right?

sicong-li-arm replied:
Done

advaitjain commented on tensorflow/lite/c/common.h:
```diff
@@ -225,6 +225,16 @@ void TfLiteFloatArrayFree(TfLiteFloatArray* a);
+    auto delta = ((a) > (b)) ? ((a) - (b)) : ((b) - (a));              \
+    if (delta > epsilon) {                                             \
+      TF_LITE_KERNEL_LOG((context), "%s:%d %s not near %s (%d != %d)", \
+                         __FILE__, __LINE__, #a, #b, (a), (b));        \
```
(a) -> static_cast<double>(a)
(b) -> static_cast<double>(b)

once you change the specifier to be %f, it will be expected to be a double. adding a static_cast in the macro definition will allow a caller to pass in floats without an implicit conversion to double error.

sicong-li-arm replied:
Done

advaitjain commented on tensorflow/lite/micro/testing/micro_test.h:
```diff
@@ -149,13 +149,26 @@ extern tflite::ErrorReporter* reporter;
+
+#define TF_LITE_MICRO_EXPECT_NEAR_RELATIVE(x, y, rel_epsilon)                  \
+  do {                                                                         \
+    auto vx = static_cast<float>(x);                                                             \
```
same comment about whether static_cast<float> can be avoided or not. Please let me know if I have missed something.

sicong-li-arm replied:
Done

advaitjain reviewed (changes requested):
thanks for the changes.

I have a few clarifying questions/comments then I think we're ready to go.

I'm only reviewing for the double promotion issue since everything else was already reviewed by @hajuho and @petewarden

advaitjain reviewed (dismissed):
Thanks for the changes. LGTM.

gbaned commented:
@sicong-li-arm Can you please resolve conflicts? Thanks!

sicong-li-arm commented:
Hi @gbaned! Resolved the conflicts and (force) rebased. Thanks! :)

sicong-li-arm commented:
Hi @gbaned @advaitjain ! All checks have passed! Could you please have a look to see if this PR can be merged? Thanks! :)

rthadur commented:
@sicong-li-arm can you please check below errors 

`third_party/tensorflow/lite/micro/kernels/softmax.cc:153:66: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
    gen_lut([](float value) { return std::exp(value); }, -10.0f, 0.0f,

third_party/tensorflow/lite/micro/kernels/softmax.cc:153:58: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
    gen_lut([](float value) { return std::exp(value); }, -10.0f, 0.0f,

third_party/tensorflow/lite/micro/kernels/softmax.cc:156:70: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
    gen_lut([](float value) { return 1.0f / (1.0f + value); }, 0.0f, 1.0f,

third_party/tensorflow/lite/micro/kernels/softmax.cc:156:64: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
    gen_lut([](float value) { return 1.0f / (1.0f + value); }, 0.0f, 1.0f,
    ~~~~~~~                                                    ^~~~
4 errors generated.`

sicong-li-arm commented:
> @sicong-li-arm can you please check below errors
> 
> `third_party/tensorflow/lite/micro/kernels/softmax.cc:153:66: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> gen_lut([](float value) { return std::exp(value); }, -10.0f, 0.0f,
> 
> third_party/tensorflow/lite/micro/kernels/softmax.cc:153:58: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> gen_lut([](float value) { return std::exp(value); }, -10.0f, 0.0f,
> 
> third_party/tensorflow/lite/micro/kernels/softmax.cc:156:70: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> gen_lut([](float value) { return 1.0f / (1.0f + value); }, 0.0f, 1.0f,
> 
> third_party/tensorflow/lite/micro/kernels/softmax.cc:156:64: error: implicit conversion increases floating-point precision: 'float' to 'double' [-Werror,-Wdouble-promotion]
> gen_lut([](float value) { return 1.0f / (1.0f + value); }, 0.0f, 1.0f,
> ~~~~~~~ ^~~~
> 4 errors generated.`

Hi @rthadur ! gen_lut expected double values instead of floats, hence the error.  @petewarden This means this will need to be reverted back to double. Would this be acceptable? An alternative would be to copy the gen_lut function over but with float computation.

I'll resolve the conflicts and push for a fix as soon as @petewarden approves. In the meantime @rthadur could you please share what tests you ran to produce this error? Somehow my local `make -f tensorflow/lite/micro/tools/make/Makefile ...` , even with double-promotion warning on failed to catch this error (maybe because it's invoking a method from lite instead of micro?).

petewarden commented:
Could you add a float variant of the gen_lut function? I'd imagine that will be a lot faster on the Cortex M cores, since they only support single precision. Thanks!

sicong-li-arm commented:
Hi @petewarden @rthadur. I resolved the conflicts and pushed a fix. Could you please have a look when you are available? Thanks!

gbaned commented:
@sicong-li-arm  Still, conflicts are appearing. Can you please resolve those? Thanks!

sicong-li-arm commented:
Hi @gbaned @rthadur, the conflict is resolved. Thanks!

gbaned commented:
@sicong-li-arm Still, conflicts are appearing. Can you please resolve those? Thanks!

sicong-li-arm commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -73,6 +101,71 @@ TfLiteStatus CalculateSoftmaxParams(TfLiteContext* context,
+    return nullptr;
+  }
+  SoftmaxOpData* op_data = static_cast<SoftmaxOpData*>(raw_op_data);
+  const TfLiteType input_type = context->tensors[0].type;
```
Hi @petewarden. The micro example tests segfaulted on this line. It turns out that micro interpreter calls the registration.init(...) as part of its AllocateTensors method (https://github.com/tensorflow/tensorflow/blob/8c6f2d55762f3fc94f98fdd8b3c5d59ee1276dba/tensorflow/lite/micro/micro_interpreter.cc#L275). However the interpreter doesn't seem to populate the context with input and output tensors at all (instead it exposes them as inputs(index) and outputs(index) methods only. Could you please advise how this can be resolved? We can either take the easy route that allocates LUTs regardless types (memory waste), or change the interpreter so that it populates the context with input+output tensors so that we can obtain the type info within init() here, although I don't know the implications of the latter approach.

petewarden replied:
@njeffrie it looks like this is related to the EvalTensors changes?

sicong-li-arm replied:
Mark this thread as resolved as the issue continues in the discussion below

sicong-li-arm reviewed (commented):
@gbaned  Thanks I'll resolve the conflicts but first I just discovered some problems that I'll need some help from @petewarden with.

gbaned commented:
@petewarden  Can you please assist on above comments from @sicong-li-arm. Thanks!

sicong-li-arm commented:
Hi @petewarden @gbaned @njeffrie  ! Sorry for leaving this hanging as I was occupied with something else. I'd like to finish this asap, so I'll try my best to respond to this PR every day.

Since the commit history becomes stale with each forced push, I'll provide a summary of the changes and remaining issues from the last push:

1. Rebased on top of @njeffrie 's latest changes with the new test approach (the new approach looks much nicer btw) and added 4 new tests for int16 in accordance with the new approach.

2. Removed redundant test helper methods (Dequantize, F2Q16S, CreateQuantizedTensor, etc.), as a result of the previous bullet.

3. (With question) Increased the tolerance for 3D and 4D int16 tests from 1 to 7. I wonder if such a tolerance is acceptable? I reckon this inaccuracy mainly stems from the conversion of the floating golden outputs to int16. Alternatively we can have another golden outputs in int16 directly, and test with bit-exactness.

4. (With question) Following [this discussion](https://github.com/tensorflow/tensorflow/pull/38873#discussion_r460820208,) above, and some suggestions from others, I tried allocating the user_data in prepare instead of init, in order to conditionally allocate based on the tensor type. It passed the kernel tests but since user_data is supposed to be allocated by init, I wonder if this approach is the right way forward? 

Thanks in advance!

sicong-li-arm commented:
Hi @petewarden @gbaned @njeffrie could you please have a look at the new changes and the remaining questions? Thanks ! :)

petewarden commented:
@njeffrie could you take a look if you get a chance? Thanks!

sicong-li-arm commented:
Hi @njeffrie could you please have a look at the new changes and the questions? Thanks!

njeffrie commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -105,44 +133,88 @@ void SoftmaxQuantized(const TfLiteEvalTensor* input, TfLiteEvalTensor* output,
-  SoftmaxParams* data = static_cast<SoftmaxParams*>(node->user_data);
-  return CalculateSoftmaxParams(context, input, output, params, data);
+  // Allocate user_data unless it's already allocated
+  if (node->user_data == nullptr) {
```
Can we get the softmax params from node->user_data like we do in other ops? I prefer having init allocate user data and prepare fill it, since that's the pattern we use elsewhere.

sicong-li-arm replied:
Done

njeffrie commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -30,23 +30,38 @@ namespace micro {
 
+// Softmax parameter data that persists in user_data
+static constexpr int size_of_lut = 513;
+struct SoftmaxOpData {
```
Can you use exp_lut and one_over_one_plus_x_lut from SoftmaxParams instead of creating this new wrapper struct? See the implementation in micro/kernels/xtensa_hifimini/softmax.cc for reference.

sicong-li-arm replied:
Done.
Yeah this wrapper struct is quite redundant. :)

njeffrie commented on tensorflow/lite/micro/kernels/softmax.cc:
```diff
@@ -30,23 +30,38 @@ namespace micro {
 namespace {
 
+// Softmax parameter data that persists in user_data
+static constexpr int size_of_lut = 513;
```
s/size_of_lut/kInt16LUTArraySize.

sicong-li-arm replied:
Done.

njeffrie reviewed (commented):
Overall the direction makes sense. It would be nice to re-use the int16 LUT pointers within SoftmaxParams instead of re-creating them if possible. This would have the side benefit of reducing the number of lines changed in micro/kernels/softmax.cc

sicong-li-arm committed:
Clean up: rename size_of_lut; remove redundant SoftmaxOpData struct

sicong-li-arm committed:
Move SoftmaxParam allocation to init; Add comments to int16 test tolerance

sicong-li-arm commented:
Hi @njeffrie @advaitjain could you please have a look at my latest commit addressing the issues we discussed? Thanks! :)

sicong-li-arm commented:
Thank you all! :)
